cedar-policy/src/api.rs

```diff
@@ -3157,6 +3157,11 @@ impl Policy {
         Ok::<_, PolicyToJsonError>(json)
     }
 
+    /// returns the json representation of a policy but strongly typed
+    pub fn to_json_policy(&self) -> Result<cedar_policy_core::est::Policy, impl std::error::Error> {
+        self.lossless.est()
+    }
```

Comment:
Since `cedar_policy_core::est::Policy` is an "internal" datatype, it is subject to change. We did not design it to be particularly pretty -- it's primarily designed to allow nice (de)serialization from JSON. 

The stable interface is the JSON policy format itself. So, if possible, I think you might be better off using the JSON directly, and then performing conversion to an EST internally to check that it has the correct form.

Reply:
I really don't think users would be better off using JSON format directly and not having a strongly typed option. But if that's what we need to do for now to unblock it, then so be it. However I would like to present my case a little better before i disagree and commit.

Remember that with wasm, people aren't actually using the `cedar_policy_core::est::Policy` type directly, so it's not quite the same as exposing it in `cedar-policy`. What people use is the typescript type generated from `cedar_policy_core::est::Policy`. In other words, any methods that might be in `cedar_policy_core::est::Policy`'s impl are not available to users of wasm. If in the future we need to break the shape of the struct, we can add a converter to the wasm package. It's not a one way door; but we do need to remember to do that if we break `cedar_policy_core::est::Policy`.

Question - isn't the format in docs.cedarpolicy.com derived from the `cedar_policy_core::est::Policy` struct shape? If not, how's it derived? If so, isn't the struct shape already exposed then? If we change the internal struct shape we'll have to create that conversion layer anyway anyway to keep backwards compatibility with the old format, or make the change during a MV update.

Reply:
The `cedar_policy_core::est::Policy` data type looks similar to the JSON, but it's not as close as you might think. For example, the data type currently [splits expressions into two classes](https://github.com/cedar-policy/cedar/blob/0333786910b491f32f80075855d1e27733f9a958/cedar-policy-core/src/est/expr.rs#L39): extension function applications and non-extension-functions. Splitting things up this way makes parsing with serde a little simpler, but this is an internal detail that it not exposed in the JSON format itself.

So: it is entirely possible that we will change the shape of `cedar_policy_core::est::Policy`s without making a change to the JSON format as described in the docs.

Reply:
If we were to let `tsify` transpile `cedar_policy_core::est::Policy` to TS, would users see the distinction between `Expr` and `ExprNoExt` (see Kesha's link)?  IMO it would be a bad user experience if users had to do something like `.expr_no_ext(...)` every time they want an expression that is not an extension function call (which is probably 99% of expressions people will create)

Reply:
Users would only see the distinction between `Expr` and `ExprNoExt` in the generated `.d.ts` type headers. The JSON itself looks like this for a policy with `Clause`/`Expr` types in its EST:
`permit(principal, action, resource) when { principal has "Email" && principal.Email == "a@a.com" };`
```
{
    "policy": {
        "effect": "permit",
        "principal": {
            "op": "All"
        },
        "action": {
            "op": "All"
        },
        "resource": {
            "op": "All"
        },
        "conditions": [
            {
                "kind": "when",
                "body": {
                    "&&": {
                        "left": {
                            "has": {
                                "left": {
                                    "Var": "principal"
                                },
                                "attr": "Email"
                            }
                        },
                        "right": {
                            "==": {
                                "left": {
                                    ".": {
                                        "left": {
                                            "Var": "principal"
                                        },
                                        "attr": "Email"
                                    }
                                },
                                "right": {
                                    "Value": "a@a.com"
                                }
                            }
                        }
                    }
                }
            }
        ]
    }
}
```